ipaclient/plugins/vault.py

```diff
@@ -928,29 +1015,51 @@ def forward(self, *args, **options):
         # initialize NSS database
         nss.nss_init(api.env.nss_dir)
 
-        # retrieve transport certificate
-        config = self.api.Command.vaultconfig_show()['result']
-        transport_cert_der = config['transport_cert']
-        nss_transport_cert = nss.Certificate(transport_cert_der)
-
         # generate session key
         mechanism = nss.CKM_DES3_CBC_PAD
         slot = nss.get_best_slot(mechanism)
         key_length = slot.get_best_key_length(mechanism)
         session_key = slot.key_gen(mechanism, None, key_length)
 
-        # wrap session key with transport certificate
-        # pylint: disable=no-member
-        public_key = nss_transport_cert.subject_public_key_info.public_key
-        # pylint: enable=no-member
-        wrapped_session_key = nss.pub_wrap_sym_key(mechanism,
-                                                   public_key,
-                                                   session_key)
+        for retry in (True, False):
+            # retrieve transport certificate
+            try:
+                transport_cert = get_cached_transport_cert(api.env.domain)
+            except KeyError:
+                config = self.api.Command.vaultconfig_show()['result']
+                transport_cert_der = config['transport_cert']
+                transport_cert = (
+                    x509.load_certificate(transport_cert_der, x509.DER))
+                set_cached_transport_cert(api.env.domain, transport_cert)
+            else:
+                transport_cert_der = (
+                    transport_cert.public_bytes(serialization.Encoding.DER))
+            nss_transport_cert = nss.Certificate(transport_cert_der)
+
+            # wrap session key with transport certificate
+            # pylint: disable=no-member
+            public_key = nss_transport_cert.subject_public_key_info.public_key
+            # pylint: enable=no-member
+            wrapped_session_key = nss.pub_wrap_sym_key(mechanism,
+                                                       public_key,
+                                                       session_key)
 
-        # send retrieval request to server
-        options['session_key'] = wrapped_session_key.data
+            # send retrieval request to server
+            options['session_key'] = wrapped_session_key.data
 
-        response = self.api.Command.vault_retrieve_internal(*args, **options)
+            try:
+                response = self.api.Command.vault_retrieve_internal(*args,
+                                                                    **options)
+            except errors.NotFound:
+                raise
+            except (errors.InternalError,
+                    errors.ExecutionError,
+                    errors.GenericError):
+                del_cached_transport_cert(api.env.domain)
```

Comment:
The cache is deleted only in case of an unexpected error, which is why there is a list of them, not in case of any error.

The KRA bug should definitely be fixed, but we should still cache the cert properly even for older servers.